### ISSUE TITLE
Decide which Subject classes should be extensible and which of their methods should be |final|

### ISSUE THREAD

**Issue by cpovirk** [google/truth#217]
Currently, we're not consistent.

As always, the easy thing to do is to lock them all down except where necessary and then open them up over time.

Some classes are final:

```
$ egrep 'final class' *Subject.java
BigDecimalSubject.java:public final class BigDecimalSubject extends ComparableSubject<BigDecimalSubject, BigDecimal> {
DoubleSubject.java:public final class DoubleSubject extends ComparableSubject<DoubleSubject, Double> {
FloatSubject.java:public final class FloatSubject extends ComparableSubject<FloatSubject, Float> {
MultisetSubject.java:public final class MultisetSubject<S extends MultisetSubject<S, E, M>, E, M extends Multiset<E>>
TableSubject.java:public final class TableSubject extends Subject<TableSubject, Table<?, ?, ?>> {
ThrowableSubject.java:public final class ThrowableSubject extends Subject<ThrowableSubject, Throwable> {
```

Some have public constructors:

```
$ egrep '^  public \w+Subject' *Subject.java
DefaultSubject.java:  public DefaultSubject(FailureStrategy failureStrategy, @Nullable Object o) {
IntegerSubject.java:  public IntegerSubject(FailureStrategy failureStrategy, @Nullable Integer integer) {
LongSubject.java:  public LongSubject(FailureStrategy failureStrategy, @Nullable Long subject) {
MultimapSubject.java:  public IterableSubject<
ObjectArraySubject.java:  public IterableSubject<?, T, List<T>> asList() {
PrimitiveBooleanArraySubject.java:  public IterableSubject<?, Boolean, List<Boolean>> asList() {
PrimitiveByteArraySubject.java:  public IterableSubject<?, Byte, List<Byte>> asList() {
PrimitiveCharArraySubject.java:  public IterableSubject<?, Character, List<Character>> asList() {
PrimitiveIntArraySubject.java:  public IterableSubject<?, Integer, List<Integer>> asList() {
PrimitiveLongArraySubject.java:  public IterableSubject<?, Long, List<Long>> asList() {
PrimitiveShortArraySubject.java:  public IterableSubject<?, Short, List<Short>> asList() {
StringSubject.java:  public StringSubject(FailureStrategy failureStrategy, @Nullable String string) {
ThrowableSubject.java:  public ThrowableSubject(FailureStrategy failureStrategy, @Nullable Throwable throwable) {
```

Some have protected constructors:

```
$ egrep '^  protected \w+Subject' *Subject.java
ComparableSubject.java:  protected ComparableSubject(FailureStrategy failureStrategy, @Nullable T subject) {
IterableSubject.java:  protected IterableSubject(FailureStrategy failureStrategy, @Nullable C list) {
```

Others have package-private constructors.

**Comment by kluever** [comment]
As a first step, all of the constructors should probably be made protected. There's really no good reason to be instantiating these AFAICT.

**Comment by cgruber** [comment]
I'm pretty ok with making at least the constructors protected, though it looks like we shouldn't likely make them non-extensible.  Too many people are already extending them (including us) and I don't think we could consistently do it.

**Comment by cgruber** [comment]
Violent agreement, kurt. :)

**Comment by cgruber** [comment]
A fair bit of this has been done - need to see what's left.  We haven't been linking to this issue in our commits, but I think we're close to done this one.

**Comment by cpovirk** [comment]
Dumping notes:

- describe these use cases and how they’d work today
  - add things to StringSubject that are about Strings
  - extend FooSubject to create SubFooSubject

**Comment by cpovirk** [comment]
Another edge case in which someone might want to subclass a `Subject`: [to create a shorter failure message](https://stackoverflow.com/a/55029785/28465) (though of course we've discussed alternative ways to accomplish this, as noted there).

**Comment by cpovirk** [comment]
For 1.0-blocking purposes, I think we've done enough here: We've made `Subject` hierarchies generally a sane thing to do.

We _could_ go back and mark various methods and classes as `final`, but I'm not sure it's worth the trouble that it would cause users who overrode them.

I did verify that visible constructors are all `protected`.

We could also go in and preemptively open up all types to subclassing. That could be a nice thing to do, but I don't think it's a 1.0 blocker, so I've filed https://github.com/google/truth/issues/577
